src/DataBase/GoalDatabase.cs

```diff
@@ -0,0 +1,162 @@
+using System.Linq;
+using ePiggyWeb.DataBase.Models;
+using ePiggyWeb.DataManagement.Entries;
+using ePiggyWeb.DataManagement.Goals;
+using ePiggyWeb.Utilities;
+
+namespace ePiggyWeb.DataBase
+{
+    public static class GoalDatabase
+    {
+        public static int Create(IGoal goal, int userid)
+        {
+            using var db = new DatabaseContext();
+            var dbGoal = new GoalModel (goal, userid);
+            db.Add(dbGoal);
+            db.SaveChanges();
+            return dbGoal.Id;
+        }
+
+        public static bool CreateList(IGoalList goalList, int userid)
+        {
+            using var db = new DatabaseContext();
+            var dictionary = new Dictionary<IGoal, IGoalModel>();
+            foreach (var goal in goalList)
+            {
+                var dbGoal = new GoalModel(goal, userid);
+                dictionary.Add(goal, dbGoal);
+            }
+            // Setting all of the ID's to local Entries, just so this method remains usable both with local and only database usage
+            db.AddRange(dictionary.Values);
+            db.SaveChanges();
+            goalList.Clear();
+            foreach (var (key, value) in dictionary)
+            {
+                key.Id = value.Id;
+                goalList.Add(key);
+            }
+            return true;
+        }
+
+        public static bool Update(IGoal oldGoal, IGoal newGoal)
+        {
+            return Update(oldGoal.Id, oldGoal.UserId, newGoal);
+        }
+
+        public static bool Update(int id, int userId, IGoal newGoal)
+        {
+            using var db = new DatabaseContext();
+            var dbGoal = db.Goals.FirstOrDefault(x => x.Id == id && x.UserId == userId);
+            if (dbGoal == null)
+            {
+                ExceptionHandler.Log("Couldn't find goal id: " + id + " in database");
+                return false;
+            }
+            dbGoal.Edit(newGoal);
+            db.SaveChanges();
+            return true;
+        }
+
+        public static bool Delete(IGoal goal)
+        {
+            return Delete(goal.Id, goal.UserId);
+        }
+
+        public static bool Delete(int id, int userId)
+        {
+            return Delete(x => x.Id == id && x.UserId == userId);
+        }
+
+        public static bool Delete(Func<GoalModel, bool> filter)
+        {
+            using var db = new DatabaseContext();
+            try
+            {
+                var dbGoal = db.Goals.FirstOrDefault(filter);
+                db.Goals.Remove(dbGoal ?? throw new InvalidOperationException());
+                db.SaveChanges();
+            }
+            catch (InvalidOperationException ex)
+            {
+                ExceptionHandler.Log(ex.ToString());
+                return false;
+            }
+            return true;
+        }
+
+        public static bool DeleteList(IEnumerable<IGoal> goalList)
+        {
+            var enumerable = goalList as IGoal[] ?? goalList.ToArray();
```

Comment:
Can you explain this line?

Reply:
First I'm trying to cast the passed goalList parameter as an array of IGoals, but if it can't be cast then instead an array of IGoals is created (not cast). It's a bit confusing, but this was resharpers suggestion and it does make sense if you think about it (thanks for the question, actually investigated it and now I understand better myself :D)